File: modules/surge-docs/src/test/scala/docs/command/BankAccountCommandModel.scala
// Copyright © 2017-2021 UKG Inc. <https://www.ukg.com>

package docs.command

import java.util.UUID

import play.api.libs.json.{ Format, JsValue, Json }
import surge.scaladsl.command.AggregateCommandModel

import scala.util.{ Failure, Success, Try }

class InsufficientFundsException(accountNumber: UUID) extends RuntimeException(s"Insufficient Funds in account $accountNumber to complete this transaction")
class AccountDoesNotExistException(accountNumber: UUID)
    extends RuntimeException(s"Account with id $accountNumber does not exist")

// #aggregate_class
object BankAccount {
  implicit val format: Format[BankAccount] = Json.format
}
case class BankAccount(accountNumber: UUID, accountOwner: String, securityCode: String, balance: Double)
// #aggregate_class

// #command_class
sealed trait BankAccountCommand {
  def accountNumber: UUID
}
case class CreateAccount(accountNumber: UUID, accountOwner: String, securityCode: String, initialBalance: Double) extends BankAccountCommand
case class CreditAccount(accountNumber: UUID, amount: Double) extends BankAccountCommand
case class DebitAccount(accountNumber: UUID, amount: Double)
    extends BankAccountCommand
// #command_class

// #event_class
sealed trait BankAccountEvent {
  def accountNumber: UUID
  def toJson: JsValue
}
object BankAccountCreated {
  implicit val format: Format[BankAccountCreated] = Json.format
}
case class BankAccountCreated(accountNumber: UUID, accountOwner: String, securityCode: String, balance: Double) extends BankAccountEvent {
  override def toJson: JsValue = Json.toJson(this)
}

object BankAccountUpdated {
  implicit val format: Format[BankAccountUpdated] = Json.format
}
case class BankAccountUpdated(accountNumber: UUID, newBalance: Double) extends BankAccountEvent {
  override def toJson: JsValue = Json.toJson(this)
}
// #event_class

// #command_model_class
object BankAccountCommandModel extends AggregateCommandModel[BankAccount, BankAccountCommand, BankAccountEvent] {
  override def processCommand(aggregate: Option[BankAccount], command: BankAccountCommand): Try[Seq[BankAccountEvent]] = {
    command match {
      case create: CreateAccount =>
        if (aggregate.isDefined) {
          // Aggregate already exists - no need to recreate
          Success(Seq.empty)
        } else {
          Success(Seq(BankAccountCreated(create.accountNumber, create.accountOwner, create.securityCode, create.initialBalance)))
        }
      case credit: CreditAccount =>
        aggregate
          .map { existing =>
            Success(Seq(BankAccountUpdated(existing.accountNumber, existing.balance + credit.amount)))
          }
          .getOrElse(Failure(new AccountDoesNotExistException(command.accountNumber)))
      case debit: DebitAccount =>
        aggregate
          .map { existing =>
            if (existing.balance >= debit.amount) {
              Success(Seq(BankAccountUpdated(existing.accountNumber, existing.balance - debit.amount)))
            } else {
              Failure(new InsufficientFundsException(existing.accountNumber))
            }
          }
          .getOrElse(Failure(new AccountDoesNotExistException(command.accountNumber)))
    }
  }

  override def handleEvent(aggregate: Option[BankAccount], event: BankAccountEvent): Option[BankAccount] = {
    event match {
      case create: BankAccountCreated  => Some(BankAccount(create.accountNumber, create.accountOwner, create.securityCode, create.balance))
      case updated: BankAccountUpdated => aggregate.map(_.copy(balance = updated.newBalance))
    }
  }
}
// #command_model_class

File: samples/java/src/main/java/com/example/command/BankAccountCommandModel.java
package com.example.command;

import com.example.account.BankAccount;
import com.example.BankAccountCreated;
import com.example.event.BankAccountEvent;
import com.example.event.BankAccountUpdated;
import com.example.account.CreateAccount;
import com.example.account.CreditAccount;
import com.example.account.DebitAccount;
import surge.javadsl.command.AggregateCommandModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class BankAccountCommandModel implements AggregateCommandModel<BankAccount, BankAccountCommand, BankAccountEvent> {
    @Override
    public List<BankAccountEvent> processCommand(Optional<BankAccount> aggregate, BankAccountCommand command) {

        if (command instanceof CreateAccount createAccount) {
            if (aggregate.isPresent()) {
                return new ArrayList<>();
            } else {
                BankAccountCreated bankAccountCreated = new BankAccountCreated(createAccount.getAccountNumber(),
                        createAccount.getAccountOwner()
                        , createAccount.getSecurityCode()
                        , createAccount.getInitialBalance());
                return Collections.singletonList(bankAccountCreated);

            }
        }
        if (command instanceof CreditAccount creditAccount) {
            if (aggregate.isPresent()) {
                BankAccount bankAccount = aggregate.get();
                BankAccountUpdated bankAccountUpdated = new BankAccountUpdated(bankAccount.accountId()
                        , bankAccount.balance() + creditAccount.getAmount());
                return Collections.singletonList(bankAccountUpdated);

            } else {
                throw new RuntimeException("Account does not exist");
            }
        }
        if (command instanceof DebitAccount debitAccount) {
            if (aggregate.isPresent()) {
                BankAccount bankAccount = aggregate.get();
                if (bankAccount.balance() >= debitAccount.getDebitAmount()) {
                    BankAccountUpdated bankAccountUpdated = new BankAccountUpdated(bankAccount.accountId(),
                            bankAccount.balance() - debitAccount.getDebitAmount());
                    return Collections.singletonList(bankAccountUpdated);

                } else {
                    throw new RuntimeException("InsufficientFund");
                }
            } else {
                throw new RuntimeException("Account does not exist");
            }
        }
        throw new RuntimeException("Invalid event command");
    }

    @Override
    public Optional<BankAccount> handleEvent(Optional<BankAccount> aggregate, BankAccountEvent event) {

        if(event instanceof BankAccountCreated bankAccountCreated){
            Optional<BankAccount> bankAccount;
             bankAccount = Optional.of(new BankAccount(event.getAccountNumber(), bankAccountCreated.getAccountOwner(),
                     bankAccountCreated.getSecurityCode(), bankAccountCreated.getBalance()));
            return  bankAccount;
        }
        if(event instanceof BankAccountUpdated bankAccountUpdated){
            return aggregate.map((item)-> new BankAccount(item.accountId(),item.accountOwner()
                    ,item.securityCode(),bankAccountUpdated.getBalance()));
        }
        throw new RuntimeException("Unhandled event");
    }

}
